src/Interpreter/instructions.cpp

```diff
@@ -71,6 +71,11 @@ Block *pop_block(Module *m) {
         m->pc_ptr = frame->ra_ptr;
     }
 
+    // free if event or proxy guard
+    if (frame->block->block_type == 0xfe || frame->block->block_type == 0xff) {
+        free(frame->block);
+    }
+
     return frame->block;
```

Comment:
We return a pointer that has been freed, seems dangerous. Are we sure we don't have any double frees?

Should we change the pointer to null to be safe? This is common practice.

Reply:
Freeing the block will not cause a double free but we should indeed set the pointer to nullptr.

However, I believe this opens a new issue. If I set the block to nullptr and consequently a nullptr is returned, the caller of `pop_block`, in this case only function `i_instr_end`, will mark the application execution as unsuccessful due to this line of code
https://github.com/TOPLLab/WARDuino/blob/ad8898624b013951dcb6a27572e84c22ab401cb4/src/Interpreter/instructions.cpp#L271

Consequently, although handling a callback or completing a proxy call would successfully remove the block on completion, the VM would still conclude that the app execution was unsuccessful. Althought, in the case of a proxy call the unsuccessful run would not break the VM since we run then in the special ProxyRun mode, the client VM would still receive an unsuccessful run status as a result of the proxy call.

I have another solution in mind, which I will first test before committing.

Reply:
Ok so a guard block is merely a marked for the end of a callback. It is in that sense a fake block. When it gets popped, Wasm should continue with the next block, since this will be the block that was executing when the callback interrupted.

In fact, I believe the pop function already returns this block immediately so the program can continue:
https://github.com/TOPLLab/WARDuino/blob/7e3b391494901e806847e85b63f4bfa9bb20e9bd/src/Interpreter/instructions.cpp#L33

So this would mean your code is freeing the wrong block.
Moreover, it passes a pointer to the freed block to the interpreter for further executing. I don't quite understand how this does not crash the VM.

Am I missing something?

Reply:
@tolauwae So the alternative that I had in mind was to move the piece of code above to become an else if statement of 
 `i_instr_end`. Through investigation, I realised that the block should be freed in the `pop_block` method as I was trying to do with my original commit. The reason is that when `pop_block` encounters a block of type callback or proxy call, it pops an additional frame from the callstack see https://github.com/TOPLLab/WARDuino/blob/ad8898624b013951dcb6a27572e84c22ab401cb4/src/Interpreter/instructions.cpp#L33

In other words, 2 frames are popped in total when dealing with callback and proxy block types. So it is thus saved to free the block of the top frame that corresponds with the proxy or callback type. However, my original code above is also not correct. The free of the block should happen before popping the next frame of the stack. commit [ef67430](https://github.com/TOPLLab/WARDuino/pull/167/commits/ef674304a9de4b730acc8c26e46cae2ce697e78d)  fixes the issue

Reply:
As I was working on the code fix and writing simultaneously an answer back on this thread, I did not notice your last reply.

So my last comment here above is still applicable. Indeed I was freeing the wrong block, which gets fixed by commit [ef67430](https://github.com/TOPLLab/WARDuino/pull/167/commits/ef674304a9de4b730acc8c26e46cae2ce697e78d). And the reason why previously the VM did not crash is that the block did never get freed because the condition of the if statement never became true.